File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Set<Integer> toBeRemoved = new HashSet<>();
        Deque<Integer> leftParens = new ArrayDeque<>();
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if(c == '(') leftParens.add(i);
            else if(c == ')') {
                if(!leftParens.isEmpty()) leftParens.removeFirst();
                else toBeRemoved.add(i);
            }
        }

        while(!leftParens.isEmpty()) toBeRemoved.add(leftParens.removeFirst());

        StringBuilder result = new StringBuilder();
        for(int i = 0; i < s.length(); i++) {
            if(!toBeRemoved.contains(i)) result.append(s.charAt(i));
        }
        return result.toString();
    }
}

File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        string result;
        unordered_set<int> toBeRemoved;
        deque<int> leftParens;
        for(int i = 0; i < s.size(); i++) {
            char c = s[i];
            if(c == '(') leftParens.push_front(i);
            else if(c == ')') {
                if(!leftParens.empty()) leftParens.pop_front();
                else toBeRemoved.insert(i);
            }
        }

        while(!leftParens.empty()) {
            toBeRemoved.insert(leftParens.front());
            leftParens.pop_front();
        }

        for(int i = 0; i < s.size(); i++) {
            if(!toBeRemoved.contains(i)) result+= s[i];
        }
        return result;
    }
};